src/setup_payload/AdditionalDataPayloadGenerator.cpp

```diff
@@ -0,0 +1,133 @@
+                                                                         size_t serialNumberBufferSize,
+                                                                         PacketBufferHandle & bufferHandle,
+                                                                         AdditionalDataFields additionalDataFields)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferTLVWriter writer;
+    TLVWriter innerWriter;
+    char rotatingDeviceIdBuffer[RotatingDeviceId::kRotatingDeviceIdHexMaxLength];
+    size_t rotatingDeviceIdBufferSize = 0;
+
+    // Initialize TLVWriter
+    writer.Init(chip::System::PacketBufferHandle::New(chip::System::PacketBuffer::kMaxSize));
+
+    err = writer.OpenContainer(AnonymousTag, kTLVType_Structure, innerWriter);
+    SuccessOrExit(err);
+
+    if (static_cast<int>(additionalDataFields & AdditionalDataFields::RotatingDeviceId) != 0)
+    {
+        // Generating Device Rotating Id
+        err = generateRotatingDeviceId(lifetimeCounter, serialNumberBuffer, serialNumberBufferSize, rotatingDeviceIdBuffer,
+                                       ArraySize(rotatingDeviceIdBuffer), rotatingDeviceIdBufferSize);
+        SuccessOrExit(err);
+
+        // Adding the rotating device id to the TLV data
+        err = innerWriter.PutString(ContextTag(kRotatingDeviceIdTag), rotatingDeviceIdBuffer,
+                                    static_cast<uint32_t>(rotatingDeviceIdBufferSize));
+        SuccessOrExit(err);
+    }
+
+    err = writer.CloseContainer(innerWriter);
+    SuccessOrExit(err);
+
+    err = writer.Finalize(&bufferHandle);
+    SuccessOrExit(err);
+
+exit:
+    return err;
+}
+
+CHIP_ERROR AdditionalDataPayloadGenerator::generateRotatingDeviceId(uint16_t lifetimeCounter, const char * serialNumberBuffer,
+                                                                    size_t serialNumberBufferSize, char rotatingDeviceIdBuffer[],
+                                                                    size_t rotatingDeviceIdBufferSize,
+                                                                    size_t & rotatingDeviceIdValueOutputSize)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    Hash_SHA256_stream hash;
+    uint8_t outputBuffer[RotatingDeviceId::kRotatingDeviceIdMaxLength];
+    uint8_t hashOutputBuffer[kSHA256_Hash_Length];
+    BufferWriter outputBufferWriter(outputBuffer, ArraySize(outputBuffer));
+    size_t rotatingDeviceIdBufferIndex                     = 0;
+    const PacketBufferHandle & lifetimeCounterBufferHandle = chip::System::PacketBufferHandle::New(2);
+    uint8_t * lifetimeCounterBuffer                        = lifetimeCounterBufferHandle->Start();
+
+    VerifyOrExit(rotatingDeviceIdBufferSize >= RotatingDeviceId::kRotatingDeviceIdHexMaxLength, err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    chip::Encoding::LittleEndian::Write16(lifetimeCounterBuffer, lifetimeCounter);
```

Comment:
One other thing.  `Write16` changes the value of `lifetimeCounterBuffer`, so later when we use that value we are putting whatever random stuff happens to be in the buffer _after_ what we put into it into the hash....

Reply:
addressed by using the temp buffer